src/shared/debug/launchConfiguration.ts

```diff
@@ -0,0 +1,64 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as _ from 'lodash'
+import * as vscode from 'vscode'
+import { AwsSamDebuggerConfiguration } from '../sam/debugger/awsSamDebugConfiguration'
+import { AWS_SAM_DEBUG_TYPE } from '../sam/debugger/awsSamDebugger'
+
+/**
+ * Reads and writes DebugConfigurations.
+ */
+export interface DebugConfigurationSource {
+    getDebugConfigurations(): vscode.DebugConfiguration[]
+    updateDebugConfigurations(value: vscode.DebugConfiguration[]): Promise<void>
+}
+
+/**
+ * Wraps read and write operations on launch.json.
+ */
+export class LaunchConfiguration {
+    /**
+     * Creates a Launch Configuration scoped to the given resource.
+     */
+    public constructor(
+        resource: vscode.Uri,
+        private readonly configSource: DebugConfigurationSource = new DefaultDebugConfigSource(resource)
+    ) {}
+
+    public getDebugConfigurations(): vscode.DebugConfiguration[] {
+        return this.configSource.getDebugConfigurations()
+    }
+
+    public getSamDebugConfigurations(): AwsSamDebuggerConfiguration[] {
+        return _(this.getDebugConfigurations())
+            .filter({ type: AWS_SAM_DEBUG_TYPE })
+            .map(config => config as AwsSamDebuggerConfiguration)
+            .value()
+    }
+
+    /**
+     * Adds a debug configuration to the top of the list.
+     */
+    public async addDebugConfiguration(debugConfig: vscode.DebugConfiguration): Promise<void> {
+        await this.configSource.updateDebugConfigurations([debugConfig, ...this.getDebugConfigurations()])
+    }
+}
+
+class DefaultDebugConfigSource implements DebugConfigurationSource {
```

Comment:
I do not see why this needs to be abstracted. For testing, actually write a file instead of mocking.

Reply:
I would make the same argument as above -- that this doesn't require callers to care about its existence since it is defaulted and the class is private. Again, an attempt to offer optional injection for testing.

I actually really dislike dealing with unit tests that work with the filesystem. Don't like when unit tests mess with shared resources. Granted, if you create a bunch of temp files they basically "belong" to the test. However, I'd argue the test impl for the current approach is much easier to write and follow then creating and cleaning temp files. 

Also, do tests run with their own clean workspace?

Reply:
>  actually really dislike dealing with unit tests that work with the filesystem. Don't like when unit tests mess with shared resources. Granted, if you create a bunch of temp files they basically "belong" to the test.

Tests _must_ mess with the filesystem because that is what the application does. If tests don't mess with fs then they don't test what the application does.

Reply:
Agree but I would do that in a larger-scoped functional test. I like unit tests to be small, no-side-effect (no global state), observable (mocks if necessary), and fast to run. There's plenty of different perfectly valid opinions on unit testing. Here it's even further complicated because we don't distinguish between unit and feature functionality tests right now. Lets discuss this further.